stdlib/public/runtime/CygwinPort.cpp

```diff
@@ -0,0 +1,149 @@
+
+  if (ret == FALSE) {
+    swift::fatalError("EnumProcessModules() failed");
+    return 0;
+  }
+
+  if (modules.size() * sizeof(HMODULE) < neededSize) {
+    modules.resize(neededSize / sizeof(HMODULE));
+    ret = EnumProcessModules(procHandle, modules.data(),
+                             modules.size() * sizeof(HMODULE), &neededSize);
+  }
+
+  if (ret == FALSE) {
+    swift::fatalError("EnumProcessModules() failed");
+    return 0;
+  }
+
+  for (unsigned int i = 0; i < neededSize / sizeof(HMODULE); i++) {
+    char modName[MAX_PATH];
+
+    if (!GetModuleFileNameExA(procHandle, modules[i], modName,
+                              sizeof(modName))) {
+      swift::fatalError("GetModuleFileNameExA() failed");
+    }
+
+    dl_phdr_info hdr;
+    hdr.dlpi_name = modName;
+    hdr.dlpi_addr = modules[i];
+
+    lastRet = Callback(&hdr, sizeof(hdr), data);
+    if (lastRet != 0)
+      break;
+  }
+
+  CloseHandle(procHandle);
+
+  return lastRet;
+}
+
+uint8_t *swift::_swift_getSectionDataPE(void *handle, const char *sectionName,
+                                        unsigned long *sectionSize) {
+  // In Cygwin, dlopen() returns PE/COFF image pointer.
+  // This is relying on undocumented feature of Windows API LoadLibrary().
+  unsigned char *peStart = (unsigned char *)handle;
+
+  int ntHeadersOffset = peStart[0x3C];
+
+  bool assert1 =
+      peStart[ntHeadersOffset] == 'P' && peStart[ntHeadersOffset + 1] == 'E';
+  if (assert1 == false) {
+    swift::fatalError("_swift_getSectionDataPE()'s finding PE failed");
+  }
+
+  unsigned char *coff = peStart + ntHeadersOffset + 4;
+
+  int16_t numberOfSections = *(int16_t *)(coff + 2);
+
+  // SizeOfOptionalHeader
+  int16_t sizeOfOptionalHeader = *(int16_t *)(coff + 16);
+
+  const int kCoffFileHeaderSize = 20;
+  unsigned char *sectionTableBase =
+      coff + kCoffFileHeaderSize + sizeOfOptionalHeader;
+
+  // Section Header Record
+  const int kSectionRecordSize = 40;
+
+  unsigned char *sectionHeader = sectionTableBase;
+  for (int i = 0; i < numberOfSections; i++) {
+    uint32_t virtualSize = *(uint32_t *)&sectionHeader[8];
+    uint32_t virtualAddress = *(uint32_t *)&sectionHeader[12];
+
+    char nameOfThisSection[9];
+    memcpy(nameOfThisSection, sectionHeader, 8);
+    nameOfThisSection[8] = '\0';
+
+    if (strcmp(sectionName, nameOfThisSection) == 0) {
+      *sectionSize = virtualSize;
+      return (uint8_t *)handle + virtualAddress;
+    }
+    sectionHeader += kSectionRecordSize;
+  }
+
+  return nullptr;
+}
+
+void swift::_swift_once_f(unsigned long *predicate, void *context,
+                          void (*function)(void *)) {
+  // FIXME: This implementation does a global lock, which is much worse than
+  // what we have on other platforms. Each swift_once should synchronize on the
+  // token.
+  swiftOnceMutex.lock();
+  if (*predicate == 0) {
+    *predicate = 1ul;
+    swiftOnceMutex.unlock();
+
+    function(context);
+  } else
+    swiftOnceMutex.unlock();
+}
```

Comment:
Does it mean that we can use ```std::once_flag``` of two words size in Cygwin, if IRGen is changed together ?
(The problem can be ignored that clang with Cygwin runtime library need a patch to use the ```std::once_flag```. I think it will be solved in the future or there is a workaround.)